grails-app/controllers/org/pih/warehouse/notification/NotificationController.groovy

```diff
@@ -104,63 +90,206 @@ class NotificationController {
+            }
+            log.info "Order subscribeUrl::${subscribeUrl}"
+            SubscribeRequest subscribeRequest = new SubscribeRequest(ORDER_TOPIC_ARN, "https", subscribeUrl)
+            subscribeRequest.returnSubscriptionArn = true
+            SubscribeResult result = notificationService.amazonSnsClient.subscribe(subscribeRequest);
+            log.info("Subscription ARN is " + result.subscriptionArn);
+        } catch (Exception e) {
+            log.error(e.printStackTrace());
+        }
+        render "subscribed"
+    }
+
+    def createOrder = {
+        JSONObject json = request.getJSON() as JSONObject
+        log.info "Create order JSON:${json}"
+        if (notificationService.confirmSubscription(json, Constants.ARN_TOPIC_ORDER_CREATE_TYPE)){
+            render "subscribed"
+            return
+        }
+        String message = json.getString("Message")
+        JSONObject oJson = new JSONObject(message)
+//        JSONArray orders = productsJson.getJSONArray("orders")
+//        for(int i = 0;i<orders.length();i++){
+//            JSONObject oJson = orders.getJSONObject(i)
+        log.info "ORDER JSON ${oJson}"
+        Requisition requisition = new Requisition()
+        if (oJson.has("name")) {
+            requisition.name = oJson.getString("name")
+        } else {
+            requisition.name = "DEFAULT"
+        }
+        try {
+            if (oJson.has("origin") && oJson.getString("origin")) {
+                JSONObject originJsonObject = oJson.getJSONObject("origin")
+                log.info "Origin location Json:${originJsonObject}"
+                Location location = Location.findByLocationNumber(originJsonObject.getString("locationNumber"))
+                if (!location && grailsApplication.config?.openboxes?.integration?.locationMapping) {
+                    log.info "Location Mapping::${grailsApplication.config?.openboxes?.integration?.locationMapping}"
+                    location = getLocationByLocationMapping(originJsonObject)
+                }
+                log.info "Origin location::${location}"
+                requisition.origin = location
             }
-            Product productInstance = Product.findByUpc(pJson.getString("id"))
-            if (!productInstance) {
-                productInstance = new Product()
-                productInstance.upc = pJson.getString("id")
+        } catch (Exception ex) {
+            log.error "Error in reading origin from order json:${oJson}"
+        }
+        try {
+            if (oJson.has("destination") && oJson.getString("destination")) {
+                JSONObject destinationJsonObject = oJson.getJSONObject("destination")
+                log.info "Destination location Json:${destinationJsonObject}"
+                Location location = Location.findByLocationNumber(destinationJsonObject.getString("locationNumber"))
+                if (!location && grailsApplication.config?.openboxes?.integration?.locationMapping) {
```

Comment:
Let's get rid of the mapping now since the new messages should include the location number. We can throw an IllegalArgumentException if the locationNumber is empty or does not match a location.